File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque


class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        dr, dc = (0, 0, 1, -1), (1, -1, 0, 0)
        n, m = len(grid), len(grid[0])
        visited = [[False for i in range(m)] for j in range(n)]
        que = deque()
        result = 0

        for i in range(n):
            for j in range(m):
                if grid[i][j] == '1' and not visited[i][j]:
                    visited[i][j] = True
                    que.append((i, j))

                    while que:
                        r, c = que.popleft()

                        for k in range(4):
                            nr, nc = r + dr[k], c + dc[k]
                            if (0 <= nr < n and
                                0 <= nc < m and
                                not visited[nr][nc] and
                                grid[nr][nc] == '1'
                                ):
                                visited[nr][nc] = True
                                que.append((nr, nc))

                    result += 1

        return result

File: 0200-number-of-islands/0200-number-of-islands.java
class Node {
    public int r;
    public int c;

    public Node(int r, int c) {
        this.r = r;
        this.c = c;
    }
}

class Solution {
    public int numIslands(char[][] grid) {
        int[] dr = {0, 0, -1, 1};
        int[] dc = {1, -1, 0, 0};
        int m = grid[0].length;
        int n = grid.length;
        boolean[][] visited = new boolean[n][m];
        Queue<Node> que = new ArrayDeque<>();
        int result = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (grid[i][j] == '1' && !visited[i][j]) {
                    visited[i][j] = true;
                    que.add(new Node(i, j));

                    while (!que.isEmpty()) {
                        Node now = que.poll();

                        for (int k = 0; k < 4; k++) {
                            int nr = now.r + dr[k];
                            int nc = now.c + dc[k];
                            if ((0 <= nr && nr < n) && (0 <= nc && nc < m) &&
                                (!visited[nr][nc]) && (grid[nr][nc] == '1')) {
                                visited[nr][nc] = true;
                                que.add(new Node(nr, nc));
                            }
                        }
                    }
                    result++;
                }
            }
        }
        return result;
    }
}